src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs

```diff
@@ -88,16 +93,24 @@ public bool Accept(string value)
             UriState = SegmentState.Inside;
 
             _lastValueOffset = _uri.Length;
-            // Allow the first segment to have a leading slash.
-            // This prevents the leading slash from PathString segments from being encoded.
-            if (_uri.Length == 0 && value.Length > 0 && value[0] == '/')
+
+            if (encode)
             {
-                _uri.Append("/");
-                _urlEncoder.Encode(Writer, value, 1, value.Length - 1);
+                // Allow the first segment to have a leading slash.
+                // This prevents the leading slash from PathString segments from being encoded.
+                if (_uri.Length == 0 && value.Length > 0 && value[0] == '/')
+                {
+                    _uri.Append("/");
+                    _urlEncoder.Encode(Writer, value, 1, value.Length - 1);
+                }
+                else
+                {
+                    _urlEncoder.Encode(Writer, value);
+                }
             }
             else
             {
-                _urlEncoder.Encode(Writer, value);
+                _uri.Append(value);
```

Comment:
We still need to encode characters other than `/`

Reply:
I'm going to guess there is no built in way to exclude encoding character and we don't want to write our own encode method or Barry will smite us.

Split value by `/`, encode each part to the writer, separated by `/`?